include/vcpkg/base/diagnostics.h

```diff
@@ -0,0 +1,271 @@
+            : m_kind(kind), m_origin(), m_position(), m_message(std::forward<MessageLike>(message))
+        {
+        }
+
+        template<class MessageLike, std::enable_if_t<std::is_convertible_v<MessageLike, LocalizedString>, int> = 0>
+        DiagnosticLine(DiagKind kind, StringView origin, MessageLike&& message)
+            : m_kind(kind), m_origin(origin.to_string()), m_position(), m_message(std::forward<MessageLike>(message))
+        {
+            if (origin.empty())
+            {
+                Checks::unreachable(VCPKG_LINE_INFO, "origin must not be empty");
+            }
+        }
+
+        template<class MessageLike, std::enable_if_t<std::is_convertible_v<MessageLike, LocalizedString>, int> = 0>
+        DiagnosticLine(DiagKind kind, StringView origin, TextRowCol position, MessageLike&& message)
+            : m_kind(kind)
+            , m_origin(origin.to_string())
+            , m_position(position)
+            , m_message(std::forward<MessageLike>(message))
+        {
+            if (origin.empty())
+            {
+                Checks::unreachable(VCPKG_LINE_INFO, "origin must not be empty");
+            }
+        }
+
+        // Prints this diagnostic to the terminal.
+        // Not thread safe: The console DiagnosticContext must apply its own synchronization.
+        void print(MessageSink& sink) const;
+        // Converts this message into a string
+        // Prefer print() if possible because it applies color
+        std::string to_string() const;
+        void to_string(std::string& target) const;
+
+        LocalizedString to_json_reader_string(const std::string& path, const LocalizedString& type) const;
+
+        DiagKind kind() const noexcept { return m_kind; }
+
+    private:
+        DiagKind m_kind;
+        Optional<std::string> m_origin;
+        TextRowCol m_position;
+        LocalizedString m_message;
+    };
+
+    struct DiagnosticContext
+    {
+        virtual void report(const DiagnosticLine& line) = 0;
+        virtual void report(DiagnosticLine&& line) { report(line); }
+
+        void report_error(const LocalizedString& message) { report(DiagnosticLine{DiagKind::Error, message}); }
+        void report_error(LocalizedString&& message) { report(DiagnosticLine{DiagKind::Error, std::move(message)}); }
+        template<VCPKG_DECL_MSG_TEMPLATE>
+        void report_error(VCPKG_DECL_MSG_ARGS)
+        {
+            LocalizedString message;
+            msg::format_to(message, VCPKG_EXPAND_MSG_ARGS);
+            this->report_error(std::move(message));
+        }
+
+    protected:
+        ~DiagnosticContext() = default;
+    };
+
+    struct BufferedDiagnosticContext final : DiagnosticContext
+    {
+        virtual void report(const DiagnosticLine& line) override;
+        virtual void report(DiagnosticLine&& line) override;
+
+        std::vector<DiagnosticLine> lines;
+
+        // Prints all diagnostics to the terminal.
+        void print(MessageSink& sink) const;
```

Comment:
```suggestion
        void print_to(MessageSink& sink) const;
```
or `print_diagnostics_to` makes it more clear what this does.